File: Neetcode Questions/Intervals/insert-interval.cpp

// https://leetcode.com/problems/insert-interval/

class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        
        vector<vector<int>> ans;
        int size = intervals.size();

        int i = 0;
        while(i < size && intervals[i][1] < newInterval[0]) ans.push_back(intervals[i++]);

        while(i < size && intervals[i][0] <= newInterval[1]){
            newInterval[0] = min(intervals[i][0], newInterval[0]);
            newInterval[1] = max(intervals[i][1], newInterval[1]);
            i++;
        }
        ans.push_back(newInterval);

        while(i < size) ans.push_back(intervals[i++]);

        return ans;
    }
};

File: Neetcode-150-js/Intervals/insert-interval.js

// https://leetcode.com/problems/insert-interval/

/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
var insert = function(intervals, newInterval) {
    let len = intervals.length;
    const ans = [];
    // if(len === 0) return ans;

    let i = 0;

    // put in all those ranges which are smaller than the new interval
    while(i < len && intervals[i][1] < newInterval[0]) ans.push(intervals[i++]);

    // encounter those ranges which fall into the interval
    while(i < len && intervals[i][0] <=  newInterval[1]){
        newInterval[0] = Math.min(newInterval[0], intervals[i][0]);
        newInterval[1] = Math.max(newInterval[1], intervals[i][1]);
        i++;
    }
    ans.push(newInterval);

    // put in all thise ranges which are greater than the new Interval
    while(i < len) ans.push(intervals[i++]);

    return ans;
};